cng/dsa_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build windows
+// +build windows
+
+package cng_test
+
+import (
+	"crypto/dsa"
+	"math/big"
+	"testing"
+
+	"github.com/microsoft/go-crypto-winnative/cng"
+	"github.com/microsoft/go-crypto-winnative/cng/bbig"
+)
+
+func TestGenerateDSAParameters(t *testing.T) {
+	testGenerateDSAParameters(t, 1024, 160)
+	testGenerateDSAParameters(t, 2048, 256)
+	testGenerateDSAParameters(t, 3072, 256)
+}
+
+func testGenerateDSAParameters(t *testing.T, L, N int) {
+	params, err := cng.GenerateDSAParameters(L)
+	if err != nil {
+		t.Errorf("%d-%d: error generating parameters: %s", L, N, err)
+		return
+	}
+
+	P := bbig.Dec(params.P)
+	Q := bbig.Dec(params.Q)
+	G := bbig.Dec(params.G)
+
+	if P.BitLen() != L {
+		t.Errorf("%d-%d: params.BitLen got:%d want:%d", L, N, P.BitLen(), L)
+	}
+
+	if Q.BitLen() != N {
+		t.Errorf("%d-%d: q.BitLen got:%d want:%d", L, N, Q.BitLen(), L)
+	}
+
+	one := new(big.Int)
+	one.SetInt64(1)
+	pm1 := new(big.Int).Sub(P, one)
+	quo, rem := new(big.Int).DivMod(pm1, Q, new(big.Int))
+	if rem.Sign() != 0 {
+		t.Errorf("%d-%d: p-1 mod q != 0", L, N)
```

Comment:
Is this logic based on an existing test, or a spec?

Reply:
This test is almost directly copied from crypto/dsa.